File: src/gui-tesing-library/Components/GTWindow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using gui_tesing_library.Controllers;
using gui_tesing_library.Interfaces;
using gui_tesing_library.Models;

namespace gui_tesing_library.Components
{
    internal class GTWindow : IGTWindow
    {
        private GTWindow_CS _windowCS;

        public GTWindow(GTWindow_CS windowCS)
        {
            _windowCS = windowCS;
        }

        public Vector2i Size
        {
            get { return new Vector2i(_windowCS.GetWindowSize()); }
        }



        public IScreenShot GetScreenshot()
        {
            return new ScreenShot(_windowCS.GetScreenshot());
        }

        public IScreenShot GetScreenshotRect(Vector2i position, Vector2i size)
        {
            return new ScreenShot(_windowCS.GetScreenshotRect(position.Native(), size.Native()));
        }

        public Color GetPixelColorAt(Vector2i postion)
        {
            return new Color(_windowCS.GetPixelColorAt(postion.Native()));
        }

        public IGTScreen PixelAtShouldBeColor(Vector2i position, Color colorColor)
        {
            _windowCS.PixelAtShouldBeColor(position.Native(), colorColor.Native());
            return this;
        }

        public Vector2i Position
        {
            get { return new Vector2i(_windowCS.GetPosition()); }
        }

        public bool DoesExist
        {
            get { return _windowCS.DoesExist(); }
        }

        public string Name
        {
            get { return _windowCS.GetWindowName(); }
        }

        public bool IsMinimized
        {
            get { return _windowCS.IsMinimized(); }
        }

        public IGTWindow MoveWindow(Vector2i offset)
        {
            _windowCS.MoveWindow(offset.Native());
            return this;
        }

        public IGTWindow Minimize()
        {
            _windowCS.Minimize();
            return this;
        }

        public IGTWindow Maximize()
        {
            _windowCS.Maximize();
            return this;
        }

        public IGTProcess GetProcessOfWindow()
        {
            return new GTProcess(_windowCS.GetProcessOfWindow());
        }

        public IGTWindow Close()
        {
            _windowCS.Close();
            return this;
        }

        public IGTWindow SetWindowSize(int x, int y)
        {
            _windowCS.SetWindowSize(x, y);
            return this;
        }

        public IGTWindow SetPostion(int x, int y)
        {
            _windowCS.SetPosition(x, y);
            return this;
        }

        public IGTWindow BringUpFront()
        {
            _windowCS.BringUpFront();
            return this;
        }

        public IGTWindow SizeShouldBe(Vector2i vector2I)
        {
            _windowCS.SizeShouldBe(vector2I.Native());
            return this;
        }

        public IGTWindow ShouldWindowExist(bool v)
        {
            _windowCS.ShouldWindowExist(v);
            return this;
        }

        public IGTWindow KillProcess()
        {
            _windowCS.KillProcess();
            return this;
        }

        public IGTWindow ShouldBeMinimized(bool state)
        {
            _windowCS.ShouldBeMinimized(state);
            return this;
        }

        public string GetWindowName()
        {
            return _windowCS.GetWindowName();
        }

        public Vector2i GetWindowContentPosition()
        {
            return new Vector2i(_windowCS.GetWindowContentPosition());
        }

        public Vector2i GetWindowContentSize()
        {
            return new Vector2i(_windowCS.GetWindowContentSize());
        }

        public Color GetContentPixelColorAt(Vector2i postion)
        {
            return new Color(_windowCS.GetContentPixelColorAt(postion.Native()));
        }

        public Color GetContentPixelColorAt(Vector2f realtivePostion)
        {
            return new Color(_windowCS.GetContentPixelColorAt(realtivePostion.Native()));
        }

        public IGTWindow ContentPixelAtShouldBeColor(Vector2i position, Color color)
        {
            _windowCS.ContentPixelAtShouldBeColor(position.Native(), color.Native());
            return this;
        }

        public IGTWindow CenterWindow()
        {
            _windowCS.CenterWindow();
            return this;
        }

        public IGTWindow WindowNameShouldBe(string title)
        {
            _windowCS.WindowNameShouldBe(title);
            return this;
        }

        public IGTWindow ContentPixelAtShouldBeColor(
            Vector2f sliderColorCheckPostion,
            Color colorshouldbe,
            int errorPass
        )
        {
            _windowCS.ContentPixelAtShouldBeColor(
                sliderColorCheckPostion.Native(),
                colorshouldbe.Native(),
                errorPass
            );
            return this;
        }

        internal GTWindow_CS native()
        {
            return this._windowCS;
        }
    }
}

File: src/GtLibrary/GTWindow.cpp
#include "pch.h"
#include "GTWindow.h"
#include "GTProcess.h"
#include <memory>
#include <string>
#include "Vector2i.h"
#include "Color.h"

#include "SystemCallsFactory.h"
#include "GTWindow.h"
#include "Helpers.h"

Vector2i GTWindow::GetPosition() const {
	return SystemCallsFactory::GetSystemCalls()->GetWindowPostion(handle);
}

bool GTWindow::DoesExist() const {
    GtConfiguration::GetInstance()->DefaultSleep();
    return SystemCallsFactory::GetSystemCalls()->DoesWindowExist(handle);
}

std::wstring GTWindow::GetName() const {
	return SystemCallsFactory::GetSystemCalls()->GetWindowName(handle);
}

bool GTWindow::IsMinimized() const {
    return SystemCallsFactory::GetSystemCalls()->IsWindowMinimized(handle);
}

void GTWindow::MoveWindow(const Vector2i& offset) {
	Vector2i currentPosition = GetPosition();
	Vector2i newPosition = currentPosition + offset;
	SystemCallsFactory::GetSystemCalls()->SetWindowPosition(handle, newPosition);
}

void GTWindow::Minimize() {
	SystemCallsFactory::GetSystemCalls()->MinizmizeWindow(handle);
    GtConfiguration::GetInstance()->DefaultSleep();
}

void GTWindow::Maximize() {
    SystemCallsFactory::GetSystemCalls()->MaximizeWindow(handle);
        GtConfiguration::GetInstance()->DefaultSleep();
}

GTProcess* GTWindow::GetProcessOfWindow() {
    return new GTProcess(SystemCallsFactory::GetSystemCalls()->GetProcessOfWindow(handle));
}

void GTWindow::Close() {
    SystemCallsFactory::GetSystemCalls()->TerminateWindow(handle);
}

void GTWindow::SetWindowSize(int x, int y) {
	Vector2i size = { x, y };
	SystemCallsFactory::GetSystemCalls()->SetWindowSize(handle, size);
}

void GTWindow::SetPosition(int x, int y) {
	Vector2i position = { x, y };
	SystemCallsFactory::GetSystemCalls()->SetWindowPosition(handle, position);
}

void GTWindow::BringUpFront() {
	SystemCallsFactory::GetSystemCalls()->BringWindowUpFront(handle);
}

void GTWindow::SizeShouldBe(const Vector2i& vector2I) {
    Helpers::ensureTrue([&]() {
        Vector2i size = SystemCallsFactory::GetSystemCalls()->GetSizeOfWindow(handle);
        return size.Equals(vector2I);
        });
}

void GTWindow::ShouldWindowExist(bool v) {
    Helpers::ensureTrue([&]() {
		bool doesExist = SystemCallsFactory::GetSystemCalls()->DoesWindowExist(handle);
        return doesExist==v;
        });
}

void GTWindow::KillProcess() {
    GTProcess* proc = this->GetProcessOfWindow();
    proc->Kill();
}

void GTWindow::ShouldBeMinimized(bool state) {
	Helpers::ensureTrue([&]() {
		bool isMinimized = SystemCallsFactory::GetSystemCalls()->IsWindowMinimized(handle);
		return isMinimized == state;
		});
}

std::wstring GTWindow::GetWindowName() const {
	return SystemCallsFactory::GetSystemCalls()->GetWindowName(handle);
}

Vector2i GTWindow::GetWindowContentPosition() const {
    Vector2i positon = this->GetPosition();
	int WindowBorderWidth = SystemCallsFactory::GetSystemCalls()->GetWindowBorderWidth();
	int WindowBorderHeight = SystemCallsFactory::GetSystemCalls()->GetWindowBorderHeight();
	int WindowTitleBarHeight = SystemCallsFactory::GetSystemCalls()->GetWindowTitleBarHeight();
    return Vector2i(
        positon.x + WindowBorderWidth,
        positon.y
        + WindowBorderHeight
        + WindowTitleBarHeight
    );
}

Vector2i GTWindow::GetWindowContentSize() const {
    Vector2i size = this->GetSize();
    int borderWidth = SystemCallsFactory::GetSystemCalls()->GetWindowBorderWidth();
    int windwopadding = SystemCallsFactory::GetSystemCalls()->GetWindowPadding();
    int WindowBorderHeight = SystemCallsFactory::GetSystemCalls()->GetWindowBorderHeight();
    int windowtielebarheight = SystemCallsFactory::GetSystemCalls()->GetWindowTitleBarHeight();
    return Vector2i(
        size.x - borderWidth * 2 - windwopadding * 2 - 1,
        size.y - WindowBorderHeight * 2 - windowtielebarheight - windwopadding * 2 - 5
    );
}

Color GTWindow::GetContentPixelColorAt(const Vector2i& position) const {
    GtConfiguration::GetInstance()->DefaultSleep();
	return SystemCallsFactory::GetSystemCalls()->GetPixelColorAt(handle, position);
}

Color GTWindow::GetContentPixelColorAt(const Vector2f& relativePosition) const {
    GtConfiguration::GetInstance()->DefaultSleep();
    Vector2i contentSize = this->GetWindowContentSize();
    return GetContentPixelColorAt(
        Vector2i(
            (int)(contentSize.x * relativePosition.x),
            (int)(contentSize.y * relativePosition.y)
        )
    );
}

void GTWindow::ContentPixelAtShouldBeColor(const Vector2i& position, const Color& color) {
	Helpers::ensureTrue([&]()->bool {
		Color pixelColor = this->GetContentPixelColorAt(position);
		return pixelColor.Equals(color);
		});
}

void GTWindow::CenterWindow() {
    GtConfiguration::GetInstance()->DefaultSleep();
    Vector2i screenSize = SystemCallsFactory::GetSystemCalls()->GetScreenSize();
    Vector2i windowSize = this->GetSize();
    Vector2i diffrence = screenSize - windowSize;
    this->SetPosition(diffrence.x / 2, diffrence.y / 2);

}

void GTWindow::WindowNameShouldBe(const std::wstring& title) {
    GtConfiguration::GetInstance()->DefaultSleep();
	Helpers::ensureTrue([&]() {
		std::wstring name = SystemCallsFactory::GetSystemCalls()->GetWindowName(handle);
		return name == title;
		});
}

void GTWindow::ContentPixelAtShouldBeColor(const Vector2f& sliderColorCheckPosition, const Color& colorShouldBe, int errorPass) {
    std::function<bool()> func = [&]()->bool {
        Color color = this->GetContentPixelColorAt(sliderColorCheckPosition);
        return color.Equals(colorShouldBe, errorPass);
        };
    Helpers::ensureTrue(func);
}

HWND GTWindow::GetHandle() {
    return this->handle;
}

Vector2i GTWindow::GetSize() const {
    return SystemCallsFactory().GetSystemCalls()->GetSizeOfWindow(this->handle);
}

GTScreenshot* GTWindow::GetScreenshot()
{
    Vector2i size = this->GetWindowContentSize();
    Vector2i pos = Vector2i();
    return   SystemCallsFactory().GetSystemCalls()->GetScreenshot(this->handle, pos, size);
}
GTScreenshot* GTWindow::GetScreenshotRect(Vector2i position, Vector2i size) {
    return   SystemCallsFactory().GetSystemCalls()->GetScreenshot(this->handle, position, size);
}

Color GTWindow::GetPixelColorAt(Vector2i position) {
    return   SystemCallsFactory().GetSystemCalls()->GetPixelColorAt(this->handle, position);
}
void GTWindow::PixelAtShouldBeColor(Vector2i position, Color color) {
    std::function<bool()> func = [&]()->bool {
        Color color = this->GetPixelColorAt(position);
        return color.Equals(color);
        };
    Helpers::ensureTrue(func);
}